src/nni_engine.hpp

```diff
@@ -252,13 +287,27 @@ class NNIEngine {
 
   NNIOperation FindNNINeighborInDAG(const NNIOperation &nni);
 
+ private:
+  // ** Access
+
+  // Get Reference of GPEngine.
+  GPEngine &GetGPEngine() { return *gp_engine_; }
+  // Get Reference of GPEngine.
+  TPEngine &GetTPEngine() { return *tp_engine_; }
+
  private:
   // Un-owned reference DAG.
   GPDAG &dag_;
   // For adding temporary NNIs to DAG.
   std::unique_ptr<GraftDAG> graft_dag_;
+  // Whether NNIEngine is using or maintaining GPEngine.
+  bool using_gp_engine_ = false;
```

Comment:
Would it make sense to use std::optional<GPEngine*> instead of the bool + GPEngine* combination?

Reply:
I was thinking I would give the option of setting a GPEngine and TPEngine, but only performing maintenance on them if they were being "used".  But I think that's probably an unnecessary complication that will rarely if ever have a use case.  Do you think optional with is better, or just use nullptr?

Reply:
I guess the simplest (nullptr) should work the best.

Reply:
Okay, changed it.